### PR TITLE
Confirm when the browser window is closed and there is unsaved work

### LINKED ISSUE #267: Confirm when the browser window is closed and there is unsaved work
This should only happen when any open documents have unsaved work.

In `main.ts` we should add `window.addEventListener("beforeunload", onWindowClose);` (see https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload) and define/export `onWindowClose` right below `onWindowResize()` in `utilities/input.rs`. It should check if any documents are currently unsaved (i.e. displaying an asterisk). If so, show the prompt with the text `"Unsaved work will be lost if the web browser tab is closed. Close anyway?"` (note: some browsers like Chrome won't display that message). This browser prompt should give the user the opportunity to leave the site or stay on the page. If there are no documents that are unsaved, this `beforeunload` event should be canceled so that the page closes immediately. (See the MDN docs for how to use this event in a cross-browser compatible manner for both the prompt and canceled cases.)

Later, when the Auto-Save feature (#233) is implemented, this should only prompt if somehow the auto-save hasn't happened yet (perhaps if writing to IndexedDB is asynchronous and a large file might take some time, or if changes are batched every few seconds to be written to IndexedDB and the save hasn't happened yet). But that's a detail to be implemented in that feature, not this one.

~~Blocked on #268~~

### PR DESCRIPTION
**Pull request by mfish33**
- browser confirmation on page exit
- prompt to save a document when closed will only trigger when unsaved

Closes #267

---
This change is [Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/397)

### CODE REVIEW

File: frontend/src/utilities/documents.ts [hunk 1]
```diff
@@ -16,22 +16,34 @@ import { panicProxy } from "@/utilities/panic-proxy";
...
+	if (targetDocument.isSaved) {
+		(await wasm).close_document(tabIndex);
+		return;
+	}
+
+	// Show the document is being prompted to close
+	await selectDocument(tabIndex);
 
-	const tabLabel = state.documents[tabIndex];
+	const tabLabel = targetDocument.name;
```
Comment (Keavon) [inline frontend/src/utilities/documents.ts:46]: Let's use `.displayName` here so we see the asterisk.

File: frontend/src/utilities/documents.ts [hunk 2]
```diff
@@ -16,22 +16,34 @@ import { panicProxy } from "@/utilities/panic-proxy";
...
 	(await wasm).select_document(tabIndex);
 }
 
 export async function closeDocumentWithConfirmation(tabIndex: number) {
-	selectDocument(tabIndex);
+	const targetDocument = state.documents[tabIndex];
+	if (targetDocument.isSaved) {
+		(await wasm).close_document(tabIndex);
+		return;
+	}
```
Comment (Keavon) [inline frontend/src/utilities/documents.ts:41]: When will this occur? As I recall the situation here, this response/message `closeDocumentWithConfirmation` should only be sent off from the backend in the case where a tab is being closed which was confirmed by the backend to be unsaved. This doesn't hurt, but it seems unnecessary. However my memory of exactly when this occurs could be spotty... does this code actually change behavior at all?

**Comment by Keavon** [comment]
Looks great!

File: frontend/src/utilities/documents.ts (hunk 2 above)
Comment (mfish33) [reply frontend/src/utilities/documents.ts:41]: closeDocumentWithConfirmation is called by the Vue frontend. The if statement stops the popup from being displayed if the document is already saved

Comment (Keavon) [reply frontend/src/utilities/documents.ts:41]: Oh, gotcha. Yes, that makes sense. I had gotten that confused with the other messages (`DisplayConfirmationToCloseDocument` and `DisplayConfirmationToCloseAllDocuments`) which is sent from the backend. This is a good chance 👍. Ready to merge!

**Review by Keavon — APPROVED** [review]

**Pull request merged**